### PR TITLE
#4459 Serialize errors in redux

### LINKED ISSUE #4459: Look into serializing full error objects in `editorSlice` and `tabStateSlice`
* They are strings right now
  * Was intended to match the formik status (string)
* Storing the full errors gives us more info/context

### PR DESCRIPTION
**Pull request by BLoe**
## What does this PR do?

- Closes #4459 
- Replaces and closes https://github.com/pixiebrix/pixiebrix-extension/pull/4526

## Discussion

@fregante Does this seem alright to you? Does this remove the need for #4526?

@twschiller @BALEHOK Anyone see anything wrong with this?

## Checklist

- [ ] Add tests
- [ ] Run Storybook and manually confirm that all stories are working
- [x] Designate a primary reviewer - @fregante

### CODE REVIEW

**Review by fregante — APPROVED** [review]
The code looks right but I will test it later today.

File: src/pageEditor/pageEditorTypes.ts [hunk 1]
```diff
@@ -93,7 +93,8 @@ export interface EditorState {
    */
   expandedRecipeId: RegistryId | null;
 
-  error: string | null;
+  // Should be ErrorObject | null but TS explodes in the slice definition (editorSlice.ts)
```
Comment (BLoe) [inline src/pageEditor/pageEditorTypes.ts:96]: You're welcome to check out the branch and make this change if you want to see what I mean. TS throws an error about depth limit on type recursion or something, on the entire `createSlice()` call.

Comment (fregante) [reply src/pageEditor/pageEditorTypes.ts]: It seems that the [JSON type that serialize-error extends](https://github.com/sindresorhus/serialize-error/blob/a212a8c3902fa1ff1fdef8f7625dd0cc6d6e89a1/index.d.ts#L11) is somehow incompatible with [immer’s WritableDraft](https://github.com/immerjs/immer/blob/ba24aab92aad7c4c4a57c87f42afcf1a898ac41f/src/types/types-external.ts#L33)

I "fixed" this by having a local version of the type without this JSON part.

**Review by fregante — APPROVED** [review]
Confirmed to fix the issue described in https://github.com/pixiebrix/pixiebrix-extension/pull/4526. I added the mention to your post so it will close it when this is merged

Screen Shot 7

File: src/pageEditor/pageEditorTypes.ts (hunk 1 above)
Comment (BLoe) [reply src/pageEditor/pageEditorTypes.ts]: We already have a type `SerializedError = Primitive | ErrorObject` defined in `core.ts` that's used in a bunch of places. It seems a little messy to introduce yet another error type, should we use `Except<>` on that `SerializedError` type to filter out the JsonObject part of it and just use that type?

File: src/errors/errorHelpers.ts [hunk 2]
```diff
@@ -334,3 +334,10 @@ export function selectError(originalError: unknown): Error {
   // Truncate error message in case it's an excessively-long JSON string
   return new Error(truncate(errorMessage, { length: 2000 }));
 }
+
+/** Basic Error-like type to be used only when ErrorObject is too complex */
```
Comment (twschiller) [inline src/errors/errorHelpers.ts:338]: What does "too complex" mean? Is this referring too the TypeScript compiler issue where it fails in type unification (e.g., due to nesting depth?)

File: src/pageEditor/EditorContent.tsx [hunk 3]
```diff
@@ -117,7 +118,7 @@ const EditorContent: React.FC = () => {
   if (editorError) {
     return (
       <div className="p-2">
-        <span className="text-danger">{editorError}</span>
+        <span className="text-danger">{getErrorMessage(editorError)}</span>
```
Comment (twschiller) [inline src/pageEditor/EditorContent.tsx:121]: 👍

File: src/errors/errorHelpers.ts (hunk 2 above)
Comment (twschiller) [reply src/errors/errorHelpers.ts:338]: Does the serialize-error library have a type we could re-use? I'm OK using this type too

File: src/pageEditor/slices/editorSlice.ts [hunk 4]
```diff
@@ -256,14 +257,15 @@ export const editorSlice = createSlice({
 
       activateElement(state, element);
     },
-    betaError(state, action: PayloadAction<{ error: string }>) {
-      state.error = action.payload.error;
+    betaError(state) {
+      const error = new BusinessError("This feature is in private beta");
```
Comment (twschiller) [inline src/pageEditor/slices/editorSlice.ts:261]: BusinessError = there's a configuration error in a user extension. This should just be an Error

File: src/pageEditor/tabState/tabStateSlice.ts [hunk 5]
```diff
@@ -122,13 +124,15 @@ export const tabStateSlice = createSlice({
...
         state.frameState = defaultFrameState;
-        state.error = error.message;
+        state.error = serializeError(error);
       })
       .addCase(awaitContextInvalidated.fulfilled, (state) => {
         state.isConnecting = false;
         state.frameState = defaultFrameState;
-        state.error =
-          "PixieBrix was updated or restarted. Reload the Page Editor to continue.";
+        const error = new BusinessError(
```
Comment (twschiller) [inline src/pageEditor/tabState/tabStateSlice.ts:132]: See comment here for more context on BusinessError: https://github.com/pixiebrix/pixiebrix-extension/pull/4590/files#r1013364031

By that definition, I agree this should be a BusinessError as it doesn't correspond to an error in our application

File: src/pageEditor/tabState/tabStateSelectors.ts [hunk 6]
```diff
@@ -16,12 +16,13 @@
...
 
 import { TabStateRootState } from "@/pageEditor/tabState/tabStateTypes";
+import { deserializeError } from "serialize-error";
 
 export const selectFrameState = ({ tabState }: TabStateRootState) =>
   tabState.frameState;
 
 export const selectTabStateError = ({ tabState }: TabStateRootState) =>
-  tabState.error;
+  tabState.error ? deserializeError(tabState.error) : null;
```
Comment (twschiller) [inline src/pageEditor/tabState/tabStateSelectors.ts:25]: 👍

**Review by twschiller — APPROVED** [review]
This works for me. We'll just have to figure out how to be careful about deserializing errors in selectors

File: src/errors/errorHelpers.ts (hunk 2 above)
Comment (BLoe) [reply src/errors/errorHelpers.ts:338]: See this thread: https://github.com/pixiebrix/pixiebrix-extension/pull/4590#discussion_r1012492614

The type coming from `serialize-error` (ErrorObject) doesn't play nicely with Immer's `WritableDraft` wrappers.

**Pull request merged**